src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -99,14 +153,38 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
                 }
                 f.to_owned()
             })
-            .ok_or(ErrorKind::ResolutionFailure)?;
-        let (_, ty_res) = cx
+            .ok_or(ErrorKind::Resolve(ResolutionFailure::NotInScope(
+                variant_name.to_string().into(),
+            )))?;
+        let ty_res = cx
             .enter_resolver(|resolver| {
                 resolver.resolve_str_path_error(DUMMY_SP, &path, TypeNS, module_id)
             })
-            .map_err(|_| ErrorKind::ResolutionFailure)?;
+            .map(|(_, res)| res)
+            .unwrap_or(Res::Err);
+        // This code only gets hit if three path segments in a row don't get resolved.
+        // It's a good time to check if _any_ parent of the path gets resolved.
+        // If so, report it and say the first which failed; if not, say the first path segment didn't resolve.
```

Comment:
Hmm, I just realized this will only run in the value namespace, so if you have a type or macro disambiguator it will still not show the first segment that failed to resolve. I should probably fix that.

Reply:
Implemented in https://github.com/rust-lang/rust/pull/75756/commits/c55f47b2081a1610559ed0ce0f028c8fb678eab1.